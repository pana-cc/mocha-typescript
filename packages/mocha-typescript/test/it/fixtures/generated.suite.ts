import { suite, test } from "../../../index";

interface TestMixin {
  [key: string]: any;
}

abstract class AbstractMixinTestBase implements TestMixin {

  [key: string]: any;
}

class TestsMixin1 {

  @test test1() {}
}

class TestsMixin2 {

  @test test2() {}
}

// courtesy of https://www.typescriptlang.org/docs/handbook/mixins.html
function applyMixins(derivedCtor: any, baseCtors: any[]) {
  baseCtors.forEach((baseCtor) => {
    Object.getOwnPropertyNames(baseCtor.prototype).forEach((name) => {
      derivedCtor.prototype[name] = baseCtor.prototype[name];
    });
  });
}

[
  { title: "google", url: "www.google.com", bases: [TestsMixin1] },
  { title: "github", url: "www.github.com", bases: [TestsMixin1, TestsMixin2] },
].forEach(({title, url, bases}) => {
  class GeneratedTestClass extends AbstractMixinTestBase {
    @test login() {}
    @test logout() {}
  }
  applyMixins(GeneratedTestClass, bases);
  suite(GeneratedTestClass);
});
