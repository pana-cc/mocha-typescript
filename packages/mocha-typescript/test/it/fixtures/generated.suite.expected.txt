
  GeneratedTestClass
    ✓ login
    ✓ logout
    ✓ test1

  GeneratedTestClass
    ✓ login
    ✓ logout
    ✓ test1
    ✓ test2

  7 passing
